### PR TITLE
Update the TQDM progress bar `on_train_epoch_end`

### LINKED ISSUE #10653: Progressbar showed different value vs tensorboard and myself output
## 🐛 Bug

code:
image

result:
image

We can see that the value on the progress bar is different from the output that I print. And the value on tensorboard also is the same as my output, so I guess there is something happening in the progress bar.

I only use one GPU.

### Environment

- PyTorch Lightning Version: 1.4.6
- PyTorch Version (e.g., 1.8): 1.8.1

cc @tchaton @rohitgr7 @akihironitta @carmocca @edward-io @ananthsub @kamil-kaczmarek @Raalsky @Blaizzy @SeanNaren @kaushikb11

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #10653

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @awaelchli

### CODE REVIEW

File: pytorch_lightning/callbacks/progress/tqdm_progress.py [hunk 1]
```diff
@@ -310,8 +319,8 @@ def print(
             s = sep.join(map(str, args))
             active_progress_bar.write(s, end=end, file=file, nolock=nolock)
 
-    def _should_update(self, current, total) -> bool:
-        return self.is_enabled and (current % self.refresh_rate == 0 or current == total)
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress/tqdm_progress.py:314]: I'm not yet fully understanding this fix. How are you handling the case where we need to update the bar at the end, when total is reached? the refresh rate may not align with the last batch.

Comment (awaelchli) [reply pytorch_lightning/callbacks/progress/tqdm_progress.py:314]: Is that why you implement `on_test_epoch_end` now, to cover that case? In that case I agree.

File: pytorch_lightning/callbacks/progress/tqdm_progress.py [hunk 2]
```diff
@@ -207,12 +208,10 @@ def init_test_tqdm(self) -> Tqdm:
         return bar
 
     def on_sanity_check_start(self, trainer, pl_module):
-        super().on_sanity_check_start(trainer, pl_module)
```
Comment (justusschock) [inline pytorch_lightning/callbacks/progress/tqdm_progress.py:210]: why is it necessary to remove the `super` calls?

**Review by justusschock — COMMENTED** [review]
Same as @awaelchli I'm struggling to understand why this fixes it. Could you maybe add a comment on that?

File: tests/callbacks/test_tqdm_progress_bar.py [hunk 3]
```diff
@@ -595,3 +596,64 @@ def test_tqdm_progress_bar_main_bar_resume():
...
+
+    model = MyModel()
+    pbar = MockedProgressBar()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=2,
+        limit_val_batches=2,
+        limit_test_batches=2,
+        max_epochs=2,
+        enable_model_summary=False,
```
Comment (awaelchli) [inline tests/callbacks/test_tqdm_progress_bar.py:636]: could also disable checkpointing

File: pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/progress/tqdm_progress.py:210]: The parent doesn't override these. It's not necessary, I can put them back if you want.

File: pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/progress/tqdm_progress.py:314]: So before, each `on_*_batch_end` checked whether it was the last batch and updated in that case (by passing `total` to `_should_update`)

Now, we do it in `on_*_epoch_end` always so this is not necessary

File: tests/callbacks/test_tqdm_progress_bar.py (hunk 3 above)
Comment (carmocca) [reply tests/callbacks/test_tqdm_progress_bar.py:636]:
```suggestion
        enable_model_summary=False,
        enable_checkpointing=False,
```

**Comment by awaelchli** [comment]
@carmocca Perhaps the PR should leave the linked issue open until we have the same fix for rich?

File: pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/progress/tqdm_progress.py:314]: I see, I have tests for that so this should be fine!

**Review by awaelchli — APPROVED** [review]

**Comment by carmocca** [comment]
> @carmocca Perhaps the PR should leave the linked issue open until we have the same fix for rich?

Sure, I don't how to unlink it, but we can re-open once this is merged

File: pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/progress/tqdm_progress.py:314]: Yes, the number of calls are the same, but the logger connector state won't be the correct one in the last `on_train_batch_end` vs `on_train_epoch_end`

**Review by kaushikb11 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
